062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,223 @@
+
+There is also a `replication_throttle` parameter, which is an upper bound on the bandwidth used to move replicas (bytes per second).
+This is already configurable in `Kafka.spec.cruiseControl.config`, and it also applies to all task executions, including replication factor changes.
+
+#### User tasks
+
+We can use the `user_tasks` endpoint to get the task execution result, passing the `User-Task-ID` header value returned by the `topic_configuration` response.
+
+Example:
+
+```sh
+$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
+{
+  "userTasks": [
+    {
+      "Status": "Completed",
+      "ClientIdentity": "127.0.0.1",
+      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
+      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
+      "StartMs": "1701961135978"
+    }
+  ],
+  "version": 1
+}
+```
+
+### Corner cases
+
+Here we analyze some events that may happen during the replication factor change execution, and how they are handled by the Topic Operator.
+
+#### Topic deleted
+
+If it's deleted in Kubernetes, the Topic Operator also deletes the topic in Kafka, and logs an error for the ongoing replication factor change, unless the `KafkaTopic` is paused or unmanaged.
+If it's deleted in Kafka, the Topic Operator recreates the topic in Kafka with the target replication factor, unless the `KafkaTopic` is paused or unmanaged.
+
+#### Cruise Control crashes
+
+Pending replication factor change requests are not durable, which means they are lost if Cruise Control crashes or is restarted after accepting the request.
+In that case, the Topic Operator gets an exception while polling, removes the pending change from its in-memory cache, and retries until Cruise Control comes back.
+
+#### Multiple changes for the same topic
+
+A reconciliation batch can only contain one event related to a specific topic, other events for the same topic are pushed back to the from of the queue.
+It is still possible that one or more replication factor changes for the same topic are requested while another one is already ongoing.
+In this case, the Topic Operator ignores intermediate changes, and only reconciles the last update when the ongoing change in completed.
+
+#### Topic Operator crashes
+
+In this case, there may be some pending and some ongoing replication factor changes.
+All pending and ongoing changes are recovered by the periodic reconciliation as soon as the Topic Operator restarts.
+
+## Affected/not affected projects
+
+The following components will be affected by this proposal:
+
+- Topic Operator: this component will drive the whole replication factor change process.
+- Cluster Operator: this component will be responsible to initialize the Cruise Control environment variables.
```

Comment:
I would like to see some chapter with some more details on the CO changes beyond the env. var changes. I suspect you would need to at least change the user settings in CC configuration to let the TO trigger the rebalances?

Reply:
What do you mean? There aren't really much more changes (see draft implementation). I'm using the admin user, same as in cluster rebalances.

Reply:
You should have your own user using the TO certificate -> that would be my expectation.

Reply:
Why can't we use the same admin user?

Reply:
Why do we use any users at all if they are shared between components?

Reply:
Fair enough. It's better to have a dedicated user so that we can identify requests coming from the TO.